Add error handling and logging to git service

nully0x opened the pull request:
This PR focused on adding possible errors for the git-service where possible.  This partly closes issue #2

nully0x committed:
chore: implement create repo

nully0x committed:
chore: write readme

nully0x committed:
chore: add error handling

nully0x committed:
chore(feat): add tracing and logging deps.

nully0x committed:
chore(feat): implement tracing and logging

nully0x commented:
This completes implementation for issue #2

daviroo commented on git-service/src/softserve/handlers.rs:
```diff
@@ -1,32 +1,103 @@
-    match create_user("test_user") {
-        Ok(_) => HttpResponse::Ok().body("Connection to Soft Serve successful"),
-        Err(e) => HttpResponse::InternalServerError().body(format!("Connection failed: {}", e)),
+pub fn check_repo_exists(repo_name: &str) -> Result<bool, GitServiceError> {
+    let command = "repo list";
+    execute_command(command)
+        .map(|output| {
+            let repos: Vec<&str> = output.lines().collect();
+            Ok(repos.iter().any(|&repo| repo.trim() == repo_name))
+        })
+        .map_err(|e| {
+            error!("Failed to check repo existence: {}", e);
+            GitServiceError::FailedToCheckRepoExistence(e.to_string())
+        })?
+}
+
+pub fn check_user_exists(username: &str) -> Result<bool, GitServiceError> {
+    let command = "user list";
+    execute_command(command)
+        .map(|output| {
+            let users: Vec<&str> = output.lines().collect();
+            Ok(users.iter().any(|&user| user.trim() == username))
+        })
+        .map_err(|e| {
+            error!("Failed to check user existence: {}", e);
+            GitServiceError::FailedToCheckUserExistence(e.to_string())
+        })?
+}
```
The `?` operator at the end of these functions is unnecessary and incorrect. 

The functions are already returning a `Result`, so using `?` here doesn't make sense. It's trying to unwrap a `Result` that's meant to be the return value. 

Remove the `?` and the `Ok()` wrapper in the `map()` closure to fix this. The `map_err()` chain will ensure the correct `Result` type is returned.

This change will preserve the error-handling logic while correctly adhering to the function's return type.
  ```rust
  pub fn check_repo_exists(repo_name: &str) -> Result<bool, GitServiceError> {
    let command = "repo list";
    execute_command(command)
        .map(|output| {
            let repos: Vec<&str> = output.lines().collect();
            repos.iter().any(|&repo| repo.split_whitespace().any(|word| word == repo_name))
        })
        .map_err(|e| {
            error!("Failed to check repo existence: {}", e);
            GitServiceError::FailedToCheckRepoExistence(e.to_string())
        })
}

pub fn check_user_exists(username: &str) -> Result<bool, GitServiceError> {
    let command = "user list";
    execute_command(command)
        .map(|output| {
            let users: Vec<&str> = output.lines().collect();
            users.iter().any(|&user| user.split_whitespace().any(|word| word == username))
        })
        .map_err(|e| {
            error!("Failed to check user existence: {}", e);
            GitServiceError::FailedToCheckUserExistence(e.to_string())
        })
}
  ```

daviroo commented on git-service/src/softserve/mod.rs:
```diff
@@ -1,17 +1,46 @@
+    let token_output = execute_command(&token_command)?;
+    let token = token_output.trim().to_string();
+
+    let server_url = config::get_server_url();
```
There's a minor issue with how the config module is being referenced in the create_repo function. To ensure we're using the correct config module (the one declared as a submodule in this file), we should use the self keyword to reference it.

```suggestion
    let server_url = self::config::get_server_url();
```

daviroo commented on README.md:
```diff
@@ -0,0 +1,127 @@
+Run `cp .env.example .env` to create a `.env` file.
+> Adjust these values as needed to match your configuration.
+
+### 3. Configure Nix
```
Very minor but we are missing a step 2 here.

daviroo reviewed (changes requested):
Just some minor suggestions.

nully0x committed:
chore(fix): resolve reviews

Update git-service/src/softserve/handlers.rs

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update README.md

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update README.md

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update README.md

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update README.md

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update README.md

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update git-service/src/softserve/handlers.rs

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update git-service/src/softserve/handlers.rs

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

Update git-service/src/softserve/handlers.rs

Co-authored-by: Theophilus <60826700+Extheoisah@users.noreply.github.com>

chore(doc): move git service setup to git-service folder

chore: modify infra doc

Extheoisah commented on README.md:
This readme is specific to setting up the git service. It should live inside the git service folder. The readme in the root should be one that explains the setup of the whole infra not just a subset

Extheoisah commented on README.md:
```diff
@@ -0,0 +1,148 @@
+
+### 3. Set up Soft Serve
+Install Soft Serve:
+> for development, soft serve executable have been provide in the nix setup for the project you don't need to install it separately.
```
```suggestion
For development, a soft serve executable has been provided in the nix setup for the project so that you don't have to install it separately.
```

Extheoisah commented on README.md:
```diff
@@ -0,0 +1,148 @@
+Install Soft Serve:
+> for development, soft serve executable have been provide in the nix setup for the project you don't need to install it separately.
+### 3.1. Before initializing Soft Serve
+> Before initializing Soft Serve, you need to set up the SSH key that will be used to authenticate with the server. This key should be added to the authorized keys in Soft Serve.
```
nit: no need to wrap this in a comment block
```suggestion
Before initializing Soft Serve, you need to set up the SSH key that will be used to authenticate with the server. This key should be added to the authorized keys in Soft Serve.
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+```bash
+ ssh-keygen -t ed25519 -C "your email or any identifier"
+```
+> This will generate a new SSH key pair in the default location (`~/.ssh/id_ed25519` and `~/.ssh/id_ed25519.pub`).
````
nit: no need to wrap this in a comment block

```suggestion
This will generate a new SSH key pair in the default location (`~/.ssh/id_ed25519` and `~/.ssh/id_ed25519.pub`).
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+ ssh-keygen -t ed25519 -C "your email or any identifier"
+```
+> This will generate a new SSH key pair in the default location (`~/.ssh/id_ed25519` and `~/.ssh/id_ed25519.pub`).
+> View the public key using the following command:
````
nit: no need to wrap this in a comment block

```suggestion
View the public key using the following command:
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+```bash
+cat ~/.ssh/id_ed25519.pub
+```
+> Copy the public key and add it to `SOFT_SERVE_INITIAL_ADMIN_KEYS` in your .env file.
````
nit: no need to wrap this in a comment block
```suggestion
Copy the public key and add it to `SOFT_SERVE_INITIAL_ADMIN_KEYS` in your .env file.
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+cat ~/.ssh/id_ed25519.pub
+```
+> Copy the public key and add it to `SOFT_SERVE_INITIAL_ADMIN_KEYS` in your .env file.
+> Next is to set config to use your private key for the server. Add the following to your `~/.ssh/config` file:
````
nit: no need to wrap this in a comment block
```suggestion
Next is to set config to use your private key for the server. Add the following to your `~/.ssh/config` file:
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+  IdentityFile ~/.ssh/id_ed25519
+  IdentitiesOnly yes
+```
+Initialize Soft Serve:
````
nit: add newline
```suggestion

Initialize Soft Serve:
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+```bash
+soft serve
+```
+> This will start the Soft Serve server on `localhost:23231`.
````
nit: no need to wrap this in a comment block and add newline after sentence
```suggestion
This will start the Soft Serve server on `localhost:23231`.

```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+```
+> This will start the Soft Serve server on `localhost:23231`.
+### 3.2. After initializing Soft Serve
+> After initializing Soft Serve, the default DB is SQLite.
````
We can reword this to:
```suggestion
The default DB for soft serve is SQLite.
```

Extheoisah commented on README.md:
```diff
@@ -0,0 +1,148 @@
+> This will start the Soft Serve server on `localhost:23231`.
+### 3.2. After initializing Soft Serve
+> After initializing Soft Serve, the default DB is SQLite.
+> You can change the DB to Postgres by setting the `SOFT_SERVE_POSTGRES_DB` in the `.env` file.
```
```suggestion
However, you can change the DB to Postgres by setting the `SOFT_SERVE_POSTGRES_DB` in the `.env` file. We use Postgres as our db for this project.
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+```
+
+
+## Setup git-service
````
```suggestion
## Setup Git Service
```

Extheoisah commented on README.md:
```diff
@@ -0,0 +1,148 @@
+
+## Setup git-service
+### Build and Run the Project
+In base directory of the project, run the following commands(also exports all the env variable needed for development):
```
nit
```suggestion
In base directory of the project, run the following commands (also ensure to have all the environment variables needed for development in the `.env` file. You can find a sample in the `.env.sample` file):
```

Extheoisah commented on README.md:
````diff
@@ -0,0 +1,148 @@
+soft serve
+```
+
+In another nix shell(you enter nix shell after running `nix develop` from the codebase base directory) cd into the `git-service` directory:
````
```suggestion
In another nix shell (you enter nix shell after running `nix develop` from the codebase base directory) cd into the `git-service` directory:
```

Extheoisah commented on git-service/src/softserve/handlers.rs:
```diff
@@ -1,32 +1,103 @@
+    user: web::Json<CreateUserRequest>,
+) -> Result<HttpResponse, GitServiceError> {
+    if check_user_exists(&user.username)? {
+        warn!("Attempt to create existing user: {}", user.username);
```
This log should be an error and not a warning since it returns an error
```suggestion
        error!("Attempt to create an existing user: {}", user.username);
```

Extheoisah commented on git-service/src/softserve/handlers.rs:
```diff
@@ -1,32 +1,103 @@
+    execute_command(command)
+        .map(|output| {
+            let repos: Vec<&str> = output.lines().collect();
+            repos.iter().any(|&repo| repo.trim() == repo_name)
```
You may want to convert the repo_name to lowercase before validation. We should do the same when saving the repo name in the database
```suggestion
            repos.iter().any(|&repo| repo.to_lowercase().trim() == repo_name.trim().to_lowercase())
```

Extheoisah commented on git-service/src/softserve/handlers.rs:
```diff
@@ -1,32 +1,103 @@
+    execute_command(command)
+        .map(|output| {
+            let users: Vec<&str> = output.lines().collect();
+            users.iter().any(|&user| user.trim() == username)
```
You may want to convert the username to lowercase before validation. We should do the same when saving the username in the database
```suggestion
            users.iter().any(|&user_name| user_name.to_lowercase().trim() == username.trim().to_lowercase())
```

Extheoisah commented on git-service/src/softserve/handlers.rs:
```diff
@@ -1,32 +1,103 @@
+    repo_req: web::Json<CreateRepoRequest>,
+) -> Result<HttpResponse, GitServiceError> {
+    if check_repo_exists(&repo_req.repo_name)? {
+        warn!("Attempt to create existing repo: {}", repo_req.repo_name);
```
This log should be an error and not a warning since it returns an error
```suggestion
        error!("Attempt to create an existing repo: {}", repo_req.repo_name);
```

Extheoisah reviewed (commented):
PR looks good. My reviews are mostly nits
